requests-for-comments/2022-11-02-data-exploration.md

```diff
@@ -0,0 +1,600 @@
+      "properties": [
+        {
+          "key": "created_at",
+          "value": "2022-11-14",
+          "operator": "is_date_exact",
+          "type": "group",
+          "group_type_index": 1
+        },
+        {
+          "key": "$browser",
+          "value": ["Chrome"],
+          "operator": "exact",
+          "type": "event"
+        },
+        {
+          "key": "email_opt_in",
+          "value": ["true"],
+          "operator": "exact",
+          "type": "person"
+        }
+      ]
+    }
+  ],
+  "display": "FunnelViz",
+  "new_entity": [],
+  "interval": "day",
+  "properties": [],
+  "funnel_viz_type": "steps",
+  "exclusions": []
+}
+```
+
+If I want to open "step 2 dropoff", the filter just gets these extra fields, and gets sent to a different API endpoint (`/insights/funnel` vs `/person/funnel`):
+
+```json
+{
+  "funnel_step_breakdown": [""],
+  "funnel_step": -2
+}
+```
+
+Here's a path with a global filter:
+
+```json
+{
+  "insight": "PATHS",
+  "properties": {
+    "type": "AND",
+    "values": [
+      {
+        "type": "AND",
+        "values": [
+          {
+            "key": "$anon_distinct_id",
+            "value": "e",
+            "operator": "not_icontains",
+            "type": "event"
+          }
+        ]
+      }
+    ]
+  },
+  "end_point": "http://amazingstore.com/cart",
+  "step_limit": 8,
+  "include_event_types": ["$pageview"],
+  "path_groupings": [],
+  "exclude_events": [],
+  "date_from": "-14d",
+  "date_to": null,
+  "funnel_filter": {
+    "date_from": "-14d"
+  },
+  "local_path_cleaning_filters": []
+}
+```
+
+Moving away from insights, here's a query under cohorts for "match persons who match any criteria: completed an event multiple times. $autocapture exactly 5 times in the last 30 days"
+
+```json
+{
+  "properties": {
+    "type": "OR",
+    "values": [
+      {
+        "type": "OR",
+        "values": [
+          {
+            "key": "$autocapture",
+            "type": "behavioral",
+            "value": "performed_event_multiple",
+            "negation": false,
+            "operator": "exact",
+            "event_type": "events",
+            "time_value": "30",
+            "time_interval": "day",
+            "operator_value": 5
+          }
+        ]
+      }
+    ]
+  }
+}
+```
+
+## Proposal: nestable typed filter objects
+
+Instead of a bespoke and scary monolithic filter object, I propose to split this up into **nestable typed filter objects**.
+
+Basically, I'd like to build something that can pass as an AST - Abstract Syntax Tree. This means we could convert it 
+to any other notation down the line (think HogQL perhaps?).
+
+The exact keywords, object types, etc. are all up for debate. Here's just the basic idea.
+
+Let's start with the simplest query that returns all events:
+
+```json
+{
+  "type": "events"
+}
+```
+
+```ts
+interface Node {
+  type: string;
+}
+```
+
+Adding a few filters
+
+```json
+{
+  "type": "events",
```

Comment:
As far as I understand, those presentation details would still be at the top level, with options depending on `type`

Reply:
I've considered two options.

**Option 1**

```json
{ 
  "type": "barTimeGraph", 
  "source": { 
    "type": "events", 
    "filter": { "date_from": "-7d" },
    "breakdown": "event.properties.$browser",
    "interval": "day"
  } 
}
```

In this option, there are 1) data AST nodes and 2) visualization AST nodes. Everything eventually reduces into a visualization node. The "data explorer" frontend will just get `#q={anything}` in the URL, and will render it as asked. If the query reduces into a data node, it'll be wrapped with some default visualization, e.g. `{ type: "dataTable" }`. 

All breakdowns, aggregations, and other filters would still be applied on the data node. There will not be that many data nodes to wrap (just "events" and "persons" probably, and not both by all visualizations), so all rendered visualization nodes should be smart enough to modify parameters inside their data nodes.

The drawback of this approach is that multiple visualizations that share the same data (think trends graph and the table below) will need special handling. Either caching, clever AST optimizations, or other kind of frontend magic to not make two queries. Seems like a solvable problem.

**Option 2**

The other option I could consider is:

```json
{ 
  "type": "events", 
  "filter": {}, 
  "breakdown": "...",
  "display": [
    "table", 
    ["lineChart", {"opt": "value"}], 
    "summaryTable"
  ]
}
```

In this case there's just one type of nodes, but they get a custom `display` property, that's frontend only. It can be modified without triggering a new API request. The frontend would still be smart enough to make an API call that's sure to include the right filters and breakdowns, such as a time interval.

....

I'm leaning towards option 1 at the moment. It feels more composable, and leans nicely into apps and custom components, if we choose to go down that route.